instrumentation/netty/netty-4.1/javaagent/src/main/java/io/lumigo/javaagent/instrumentation/netty/v4_1/NettyChannelPipelineInstrumentation.java

```diff
@@ -0,0 +1,133 @@
+  }
+
+  @Override
+  public void transform(TypeTransformer transformer) {
+    transformer.applyAdviceToMethod(
+        isMethod()
+            .and(nameStartsWith("add").or(named("replace")))
+            .and(takesArgument(1, String.class))
+            .and(takesArgument(2, named("io.netty.channel.ChannelHandler"))),
+        NettyChannelPipelineInstrumentation.class.getName() + "$ChannelPipelineAddAdvice");
+  }
+
+  @SuppressWarnings("unused")
+  public static class ChannelPipelineAddAdvice {
+
+    private static final Logger logger = Logger.getLogger(ChannelPipelineAddAdvice.class.getName());
+
+    @Advice.OnMethodEnter
+    public static void trackCallDepth(
+        @Advice.Argument(2) ChannelHandler handler,
+        @Advice.Local("lumigoCallDepth") CallDepth callDepth) {
+      // With Java 21 we have error from Muzzle:
+      // Missing method io.netty.channel.ChannelHandler#getClass()Ljava/lang/Class;
+      // This is probably, because the class is not loaded by the same classloader as the agent.
+      // so we can't use handler.getClass() method on it.
+      // Similar to
+      // io.opentelemetry.javaagent.instrumentation.netty.v4_1.NettyChannelPipelineInstrumentation.ChannelPipelineAddAdvice.trackCallDepth
+      //      if (callDepth == null) {
+      //        callDepth = CallDepth.forClass(ChannelPipeline.class);
+      //        callDepth.getAndIncrement();
+      //      }
+    }
+
+    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+    public static void addHandler(
+        @Advice.This ChannelPipeline pipeline,
+        @Advice.Argument(1) String handlerName,
+        @Advice.Argument(2) ChannelHandler handler,
+        @Advice.Local("lumigoCallDepth") CallDepth callDepth) {
+
+      // Here we have to check if the call depth is greater than 1, because the handler is added
+      // together with
+      // io.opentelemetry.javaagent.instrumentation.netty.v4_1.NettyChannelPipelineInstrumentation.ChannelPipelineAddAdvice.addHandler
+      //      if (callDepth.decrementAndGet() > 1) {
+      //        return;
+      //      }
+
+      VirtualField<ChannelHandler, ChannelHandler> virtualField =
+          VirtualField.find(ChannelHandler.class, ChannelHandler.class);
+
+      try {
+        // Server pipeline handlers
+        if (Objects.equals(handlerName, OPENTELEMETRY_RELOCATED_HANDLER_CLASS_NAME)) {
+
+          // if our handler is already attached, don't attach it again
+          if (virtualField.get(handler) != null) {
+            return;
+          }
+
+          ChannelHandler ourHandler = new HttpServerTracingHandler();
+          // With Java 21 we have error from Muzzle:
+          // Missing method io.netty.channel.ChannelHandler#getClass()Ljava/lang/Class;
+          // This is probably, because the class is not loaded by the same classloader as the agent
+          // so we can't use getClass() method on it.
```

Comment:
So in the end the agent crashed because of the `getClass()`?

Reply:
no, it was crashing because of Classloader issue